File: collections/index-of/scala/IndexOf.scala
object IndexOf {
  def main(args: Array[String]) {
    val elementFound = 3
    val elementNotFound = 5
    val list = List(1, 2, 3, 4)
    println(indexOf(list, elementNotFound))
    println(indexOf(list, elementFound))
  }

  def indexOf(list: List[Int], element: Int): Int = {
    def indexOfWithIndex(list: List[Int], element: Int, index: Integer): Int = list match {
      case Nil                             => -1
      case head :: Nil                     => -1
      case head :: tail if head == element => index
      case head :: tail                    => indexOfWithIndex(tail, element, index + 1)
    }

    indexOfWithIndex(list, element, 0)
  }
}

File: collections/index-of/java/IndexOf.java
import java.util.List;
import java.util.Arrays;

public class IndexOf {
	public static void main(final String... args) {
		final int elementFound = 3;
		final int elementNotFound = 5;
		final List<Integer> list = Arrays.asList(1, 2, 3, 4);
		
		System.out.println(indexOf(list, elementNotFound));
		System.out.println(indexOf(list, elementFound));
	}
	
	private static int indexOf(final List<Integer> list, final int element) {
		int index = 0;
		for (final Integer i : list) {
			if (i == element) {
				return index;
			}
			index++;
		}
		return -1;
	}
}
